### ISSUE TITLE
Однофайловые компоненты vue3

### ISSUE THREAD

**Issue by Voral** [bitrix-tools/cli#61]
Очень нужны документация и пример. Перечитал перепроверил два закрытых тикета. Прочитав их добился что вроде как начинает компилироваться, но выдает
`Error: Cannot read properties of undefined (reading 'type')`

Переустановил, как сказано, в закрытом тикете bitrix/cli  теперь вернулась
`Error: Unexpected character '#' (5:4) in ..../local/js/test/test/src/test.js`

Код следующий
bundle.config.js
`const isProd = process.argv.includes('--prod');
const vuePlugin = require('rollup-plugin-vue');
const commonjs = require('rollup-plugin-commonjs');

module.exports = {
    input: 'src/test.js',
    output: 'dist/test.bundle.js',
    namespace: 'BX.Test',
    browserslist: true,
    minification: isProd,
    sourceMaps: !isProd,
    plugins: {
        custom: [
            vuePlugin(),
            commonjs(),
        ],
    },
};`

test.js
`import {BitrixVue} from 'ui.vue3';
import Hello from "Hello.vue";

export class Test {
    #application;
    #rootNode;

    constructor(rootNode): void {
        this.#rootNode = document.querySelector(rootNode);
    }

    start(): void {
        this.#application = BitrixVue.createApp({
            name: 'Example',
            components:{
                Hello
            },
            template: '<div>Test: <Hello/></div>'
        })
        this.#application.mount(this.#rootNode);
    }
}
`

Hello.vue
`<template>
  <p class="greeting">{{ greeting }}</p>
</template>

<script>
export default {
  data() {
    return {
      greeting: 'Привет всем!'
    }
  }
}
</script>

<style>
.greeting {
  color: red;
  font-weight: bold;
}
</style>`

**Comment by Voral** [comment]
package.json
```json
{
  "dependencies": {
    "rollup-plugin-commonjs": "^9.0.0",
    "rollup-plugin-vue": "^5.0.0"
  }
}

```

**Comment by Voral** [comment]
Info @bitrix/cli, v3.2.2, так же
```
   Build error: Unexpected character '#' (5:4) in .../local/js/test/test/src/test.js
   .../local/js/test/test/src/test.js:5:4
```

**Comment by druxik** [comment]
Есть у кого-нибудь рабочий пример с однофайловыми компонентами vue3?
когда делаю import App from "./components/App.vue"; в bunle закидывается не та переменная
![image](https://github.com/bitrix-tools/cli/assets/14878717/1bd238d3-0b85-4a84-9a70-610ea3bab9a7)
Посмотрел в исходник, что пытается подключить cli
[ 'ui.vue3', 'vue' ] 
'vue' появляется, когда делаю import App from "./components/App.vue";

**Comment by Voral** [comment]
> Есть у кого-нибудь рабочий пример с однофайловыми компонентами vue3? 

Долго пытался найти решение и в поддержку обращался (ответили что не поддерживается). В итоге либо без однофайловых либо использую vite (т.е без bitrix/cli)

**Comment by druxik** [comment]
![image](https://github.com/bitrix-tools/cli/assets/14878717/2c000012-e6f7-437e-8d95-09ca244832f0)
да, тоже получил ответ от поддержки, печально(

**Comment by svn72** [comment]
> Есть у кого-нибудь рабочий пример с однофайловыми компонентами vue3?

Пропатчи себе bitrix/cli изменениями из #59 и будет тебе счастье, там дел на пару строк кода

**Comment by druxik** [comment]
> > Есть у кого-нибудь рабочий пример с однофайловыми компонентами vue3?
> 
> Пропатчи себе bitrix/cli изменениями из #59 и будет тебе счастье, там дел на пару строк кода

Как это правильно делается, никогда этого не делал)
Допустим я установил cli через npm и он находится по пути
C:\Users\Roma\AppData\Roaming\npm\node_modules\@bitrix\cli
Туда иду и делаю следующее
`
git init
git commit ....
curl https://patch-diff.githubusercontent.com/raw/bitrix-tools/cli/pull/59.patch | git am
`
Получаю Patch is empty.
Как это правильно делается? Желательно, чтобы это легко повторили другие разработчики из команды)

**Comment by svn72** [comment]
@druxik 
> Получаю Patch is empty.

Судя по документации git am требует именно файл, потому и ошибка.
У меня срабатывает следующий порядок действий (нода - v20.10)


> Туда иду и делаю следующее

```
git init
git commit ...
wget https://patch-diff.githubusercontent.com/raw/bitrix-tools/cli/pull/59.patch
git apply 59.patch
npm run build
```

Абсолютно стабильным этот алгоритм не назвать, т.к. версия в npm часто выше чем тут в репозитории.  
Так, что будем ждать.  
Возможно ПР когда-то вольется, или мейнтейнеры что-нибудь доработают по своему, и можно будет писать vue компоненты как написано в доке по vue

**Comment by Dictator90** [comment]
2025 год на дворе, а однофайловые компоненты так и не завезли?

**Comment by Dictator90** [comment]
@svn72 Да, я видел и пользовался тем костылем Vue от Bitrix. Но недавно познал всю прелесть Composition API и однофайловых компонентов. Затянул ваши правки из https://github.com/bitrix-tools/cli/pull/59 .Но не очень понимаю что это дало)) 

Попытался так 

```
const webpack = require('webpack');
const { VueLoaderPlugin } = require('vue-loader');

module.exports = {
	input: 'src/index.js',
	output: 'dist/test.bundle.js',
	globals: {
		'vue': 'createApp'
	},
	namespace: 'MB.Vue.Test',
	plugins: {
		resolve: true,
		custom: [
			new VueLoaderPlugin()
		],
	},
}
```


Но на выходе ... 
`
(function (exports,Hello) 
...
((this.MB.Vue.Test = this.MB.Vue.Test || {}),BX));
`

Видно что Vue завелся (правда ругается на все подряд) Но сама суть что компонента нет.. и он BX

**Comment by Dictator90** [comment]
@svn72 Тут был весь смысл использовать Bitrix Js Extension , а так по факту можно и просто через Vite

**Comment by Dictator90** [comment]
@svn72 Огонь! Пока не посмотрел тоже возникла идея  просто собрать на Vite а потом добавить config.php с верными путями - и js extension без проблем подключится.

**Comment by svn72** [comment]
@Dictator90 
Смысл того PR и был добавлении возможности прокидывать алиас через globals.
Чтобы при сборке в iife режиме (в котором вобщем-то и работает bitrix/cli) в итоговой функции было что-то вроде
```js
function(vue) {...}(BX.Vue3)
```
При этом сам vue тащить из ядра, т.к. общие зависимости из `.config.php` экстеншенов при подключении через 
```php
\Bitrix\Main\UI\Extension::load(...)
```
или 
```js
BX.loadExt(...);
```
повторно загружаться не будут.


А если vue будет помещен в итоговый бандл, то это может привести к избыточному его увеличению.
Не стоит забывать, что экстеншенов может быть бесконечно много.
